homeassistant/components/todo/__init__.py

```diff
@@ -489,7 +492,13 @@ async def _async_update_todo_item(entity: TodoListEntity, call: ServiceCall) ->
     if summary := call.data.get("rename"):
         updated_data["summary"] = summary
     if status := call.data.get("status"):
+        original_status = updated_data.get("status")
         updated_data["status"] = status
+        if (
+            original_status != TodoItemStatus.COMPLETED
+            and status == TodoItemStatus.COMPLETED
+        ):
+            updated_data["completed"] = dt_util.utcnow()
```

Comment:
Is this appropriate to do here in the base class, or does this need to be re-implemented (or not done at all) in every todolist provider?

Reply:
No, we should not do this here since this may be a read-only property for many implementations -- for example in the Google tasks the source of truth for this is the server.

Reply:
In the case of local todo, i think it would make sense to do it in the ical library.

We probably also want to *unset* the completed date when the entry changes status to not completed and some other logic like that.

Reply:
I agree with moving this to local_todo, but could you elaborate what role you see the ical library would play in this feature? I would have thought the integration would just set the `completed` property on the `Todo` item in `async_update_todo_item`. I didn't see an additional role for the library there.

Reply:
Similar to how it manages the `dtstamp`, `created`, and `updated` property https://github.com/allenporter/ical/blob/d6d866ed0c658172ea3155570dc2f4c1ca696537/ical/store.py#L186  it can also manage the `completed` timestamp.

There are other scenarios to consider: async_create_todo_item can also create an item in the completed state (though rare, its possible from the API) and the store can manage the full lifecycle. It can also manage stuff like the item was marked uncompleted and clear the completed timestamp when not completed if that makes sense. Note also when persisting to rfc5545 the spec states `The value MUST be specified as a date with UTC time.` and this is handled already by the `dtstamp_fn` in `ical`. This is an internally managed detail of the library.

Reply:
Just to confirm, are you confident that clearing the timestamp when uncompleting the item is the ideal behavior?

I had thought that it would possibly be informative, when you have an active item in the todo list, to still be able to check the last time it had been completed. Wiping the timestamp when clearing completed seemed to be needless destruction of information. 

But I'm not confident what would be considered expected behavior, so let me know if that is your preference.